### DISCUSSION TITLE
Outputting Resources

### DISCUSSION THREAD

**Discussion by livingstonjames45** [pulumi/pulumi#7568] [Pulumi in Practice] [answered]
With terraform, I could output individual resource properties, such as:
```
output "main_vpc_id" {
  value = module.vpc.main_vpc.id
}
```

Often, I found this cumbersome because while it probably helps terraform identify dependencies, if you have a lot of properties to return for a single resource it's a bit bulky. In those cases, I would simply return the entire vpc resource as an output, and access properties off of that.

```
output "main_vpc" {
  value = module.vpc.main_vpc
}
```

or, I could return a new object for specific properties, and just pass that object around. Here's some pseudocode:

```
output "main_vpc" {
  value = {
    id = (whatever the vpc is)
    name = vpc.name
   cidr_thing = vpc.cidr_block
   route_table = vpc.main_route_table.id
  }
}
```

Then, I would pass this around as needed, and other modules would simply access its properties.

Is this recommended in pulumi?
What's the best way of passing around a dynamic object of dynamic properties like above, or should I just use a typed class of typed properties to hold those values?

**Comment by Dagon-** [comment]
>  In those cases, I would simply return the entire vpc resource as an output, and access properties off of that.

No idea if it's the recommended way or not but this is what I do when passing outputs between values between stacks in pulumi. As you noted passing individual properties quickly becomes cumbersome.

**Comment by nimbinatus** [comment, marked as answer]
Just for completeness: Exporting an entire resource gets a bit cumbersome as it's JSON-serialized and is generally fairly large. That's why we recommend exporting each value individually, but it certainly can be done as noted above (and I completely get why it's awkward to export each value when you get into more practical builds). However, if you want to reference values across stacks, I think you're looking for [stack references](https://www.pulumi.com/docs/intro/concepts/stack/#stackreferences). Pulumi uses stack references to share values to other stacks, letting you reference the values elsewhere in your code.
